git/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+pub enum Error {
+    #[error("io error {error} at {path}")]
+    Io {
+        error: std::io::Error,
+        path: std::path::PathBuf,
+    },
+
+    #[error("target `{0}` not found")]
+    TargetNotFound(Target),
+
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+trait IoResultExt<T> {
+    fn with_path<P: AsRef<Path>>(self, path: P) -> Result<T>;
+}
+
+impl<T> IoResultExt<T> for Result<T, std::io::Error> {
+    fn with_path<P: AsRef<Path>>(self, path: P) -> Result<T> {
+        self.map_err(|error| Error::Io {
+            error,
+            path: path.as_ref().to_owned(),
+        })
+    }
+}
+
+trait InternalResultExt<T> {
+    fn wrap_err(self) -> Result<T>;
+}
+
+impl<T, E: Into<anyhow::Error>> InternalResultExt<T> for Result<T, E> {
+    fn wrap_err(self) -> Result<T> {
+        self.map_err(|e| Error::Internal(e.into()))
+    }
+}
+
+/// Specifies a git location that we can fetch.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Spec {
+    /// The url of a git repository.
+    pub url: gix::Url,
+    /// Specifies the target commit within the repository.
+    pub target: Target,
+}
+
+impl Spec {
+    /// The HEAD commit at the given url.
+    pub fn head(url: gix::Url) -> Self {
+        Self {
+            url,
+            target: Target::Head,
+        }
+    }
+
+    pub fn commit(url: gix::Url, commit: ObjectId) -> Self {
+        Self {
+            url,
+            target: Target::Commit(commit),
+        }
+    }
+}
+
+/// The different kinds of git "thing" that we can target.
+#[serde_with::serde_as]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Default, serde::Serialize, serde::Deserialize)]
+pub enum Target {
+    /// By default, we target the remote HEAD.
+    #[default]
+    Head,
+    /// Target the tip of a specific branch.
+    Branch(String),
+    /// Target a specific tag.
+    Tag(String),
+    /// Target a specific commit.
+    ///
+    /// Currently, we only support a full commit: this needs to be a full hex-encoded
+    /// sha hash. We could try to support prefixes also, but it requires some work
+    /// and it *appears* to be inherently less efficient because there doesn't seem
+    /// to be a way to fetch it in one shot. At least, when `cargo` needs to fetch an
+    /// abbreviated hash it fetches everything and then looks for the hash among the
+    /// things that it finds.
+    Commit(#[serde_as(as = "serde_with::DisplayFromStr")] ObjectId),
+}
+
+impl std::fmt::Display for Target {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Target::Head => write!(f, "HEAD"),
+            Target::Branch(branch) => write!(f, "refs/heads/{branch}"),
+            Target::Tag(tag) => write!(f, "refs/tags/{tag}"),
+            Target::Commit(c) => write!(f, "{}", c),
+        }
+    }
+}
+
+impl Target {
+    fn refspec(&self) -> String {
```

Comment:
Is it worth it to have an alias instead of just using `to_string()`? Or do you think it conveys the meaning more precisely?

Reply:
Yeah, probably not worth it. I'll remove.